nemo/collections/asr/parts/preprocessing/features.py

```diff
@@ -428,5 +476,175 @@ def forward(self, x, seq_len):
             pad_amt = x.size(-1) % pad_to
             if pad_amt != 0:
                 x = nn.functional.pad(x, (0, pad_to - pad_amt), value=self.pad_value)
-
         return x, seq_len
+
+    def _daligraph_reflect_pad(self, x, x_len, stft_pad_amount):
+        def flip_1d(x):
+            # TODO(janton): remove the layout trick when Flip supports arbitrary data layouts
+            x = dali.fn.reshape(x, shape=(-1, 1, 1), layout="HWC")
+            x = dali.fn.flip(x, vertical=1)
+            x = dali.fn.reshape(x, shape=(-1,), layout="t")
+            return x
+
+        pad_start = dali.fn.slice(x, start=1, shape=stft_pad_amount, axes=(0,))
+        pad_start = flip_1d(pad_start)
+
+        pad_end = dali.fn.slice(x, start=(x_len - stft_pad_amount - 1), shape=stft_pad_amount, axes=(0,))
+        pad_end = flip_1d(pad_end)
+        x = dali.fn.cat(pad_start, x, pad_end, axis=0)
+        return x
+
+    def _daligraph_splice_frames(self, x, nfeatures, x_len, stacking=1, subsampling=1):
+        if stacking > 1:
+            seq = [x]
+            for n in range(1, stacking):
+                f = dali.fn.slice(x, start=n, shape=x_len, axes=(1,), out_of_bounds_policy='pad', fill_values=0)
+                seq.append(f)
+            x = dali.fn.cat(*seq, axis=0)
+            nfeatures = nfeatures * stacking
+        if subsampling > 1:
+            out_len = (x_len + subsampling - 1) // subsampling
+            m = dali.fn.transforms.scale(scale=[subsampling, 1], center=[0.5, 0])
+            x = dali.fn.reshape(x, rel_shape=[1, 1, -1], layout="HWC")  # Layout required by WarpAffine
+            x = dali.fn.warp_affine(
+                x, matrix=m, size=dali.fn.cat(nfeatures, out_len), interp_type=dali.types.INTERP_NN
+            )
+            x = dali.fn.reshape(x, rel_shape=[1, 1], layout="ft")
+        return x
+
+    def init_dali_pipeline(self, batch_size, device, device_id, num_threads=4):
+        pipe = dali.pipeline.Pipeline(
```

Comment:
Will 4 threads be sufficient for larger training scenarios (slurm clusters)?

Reply:
Those are 4 CPU data-processing threads per pipeline (1 pipeline per GPU instance). In a large training scenario you probably have more than one GPU and end up with ``n_gpus * 4`` threads only for DALI data processing. From our experience, going beyond 4 threads doesn't improve training performance. Also, most of the data processing in this pipeline should happen in the GPU anyway.

Reply:
Makes sense, thanks for explanation